File: Java/PrimeNumbers.java
// JDK 23.0.1

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class PrimeNumbers {
    public static void main(String[] args) {
        final int LIMIT = 1_000_000_000;
        final String OUTPUT_FILE = "prime_numbers.txt";

        System.out.println("Finding all prime numbers up to " + LIMIT + "...");

        // Start the stopwatch
        long startTime = System.currentTimeMillis();

        // Calculate primes
        List<Integer> primes = sieveOfEratosthenes(LIMIT);

        // Stop the stopwatch
        long endTime = System.currentTimeMillis();
        double elapsedTime = (endTime - startTime) / 1000.0;

        // Display results
        System.out.println("\nNumber of primes found: " + primes.size());
        System.out.printf("Time taken: %.2f seconds%n", elapsedTime);

        // Save primes to file
        System.out.println("\nSaving primes to file...");
        savePrimesToFile(primes, OUTPUT_FILE);
        System.out.println("All primes have been saved to '" + OUTPUT_FILE + "'");
    }

    private static List<Integer> sieveOfEratosthenes(int limit) {
        boolean[] isPrime = new boolean[limit + 1];
        for (int i = 2; i <= limit; i++) {
            isPrime[i] = true;
        }

        int sqrtLimit = (int) Math.sqrt(limit);
        System.out.println("Processing primes...");

        for (int p = 2; p <= sqrtLimit; p++) {
            if (isPrime[p]) {
                for (long multiple = (long) p * p; multiple <= limit; multiple += p) {
                    isPrime[(int) multiple] = false;
                }
            }
            printProgressBar(p, sqrtLimit);
        }
        printProgressBar(1, 1); // Complete progress bar
        System.out.println();

        // Collect primes into a list
        List<Integer> primes = new ArrayList<>();
        for (int i = 2; i <= limit; i++) {
            if (isPrime[i]) {
                primes.add(i);
            }
        }
        return primes;
    }

    private static void savePrimesToFile(List<Integer> primes, String filename) {
        try (BufferedWriter writer = new BufferedWriter(new FileWriter(filename))) {
            for (int prime : primes) {
                writer.write(prime + "\n");
            }
        } catch (IOException e) {
            System.err.println("Failed to save primes: " + e.getMessage());
        }
    }

    private static void printProgressBar(int current, int total) {
        int progress = (int) ((double) current / total * 100);
        StringBuilder bar = new StringBuilder("[");
        for (int i = 0; i < 50; i++) {
            if (i < progress / 2) {
                bar.append("=");
            } else if (i == progress / 2) {
                bar.append(">");
            } else {
                bar.append(" ");
            }
        }
        bar.append("] ").append(progress).append("%\r");
        System.out.print(bar);
    }
}

File: TypeScript/src/PrimeNumbers.ts
import * as fs from 'fs';
import { SingleBar, Presets } from 'cli-progress';
import { performance } from 'perf_hooks';

// Function to find prime numbers using Sieve of Eratosthenes
const sieveOfEratosthenes = (limit: number): number[] => {
    const isPrime = new Uint8Array(limit + 1).fill(1); // Array to mark prime numbers
    isPrime[0] = 0; // 0 is not prime
    isPrime[1] = 0; // 1 is not prime

    for (let i = 2; i * i <= limit; i++) {
        if (isPrime[i]) {
            for (let j = i * i; j <= limit; j += i) {
                isPrime[j] = 0; // Mark multiples of i as non-prime
            }
        }
    }

    const primes: number[] = [];
    for (let i = 2; i <= limit; i++) {
        if (isPrime[i]) {
            primes.push(i);
        }
    }
    return primes;
};

// Main function to manage process and display progress
const main = async (): Promise<void> => {
    const limit = 1_000_000_000; // Define the limit 1 billion
    const chunkSize = 100_000; // Process data in chunks to show progress

    const progressBar = new SingleBar({}, Presets.shades_classic);

    // Start the stopwatch
    const startTime = performance.now();

    progressBar.start(limit, 0);

    const fileStream = fs.createWriteStream('prime_numbers.txt');
    let primeCount = 0;

    for (let start = 2; start <= limit; start += chunkSize) {
        const end = Math.min(start + chunkSize - 1, limit);
        const primes = sieveOfEratosthenes(end);

        // Filter primes within the chunk range
        const chunkPrimes = primes.filter(prime => prime >= start && prime <= end);
        
        // Write primes to file
        fileStream.write(chunkPrimes.join('\n') + '\n');

        primeCount += chunkPrimes.length;
        progressBar.update(start);
    }

    fileStream.end(); // Close the file stream
    progressBar.stop();

    // Stop the stopwatch
    const endTime = performance.now();
    const elapsedTime = ((endTime - startTime) / 1000).toFixed(2); // Convert to seconds

    console.log(`\nAll primes have been saved to 'prime_numbers.txt'`);
    console.log(`\nNumber of primes found: ${primeCount}`);
    console.log(`Time taken: ${elapsedTime} seconds`);
};

// Run the program
main().catch(err => console.error(err));